File: orchestrator/src/core/scope.ts
// Scope fence math. The structured engine is filesystem-aware and fail-closed;
// the lexical surface remains only for pre-canonicalized legacy CLI callers.

import { lstatSync, realpathSync, statSync } from "fs";
import { dirname, isAbsolute, join, relative, resolve, sep } from "path";

export interface ScopeInput {
  /** Declared paths for the ticket (directory prefixes or file paths). */
  declaredPaths: string[];
  /** Target path to check (already canonicalized to be worktree-root-relative). */
  targetPath: string;
  /** Whether this is a write operation. */
  isWrite: boolean;
}

export type ScopeVerdict =
  | { result: "ALLOW" }
  | { result: "DENY"; reason: string; code: string };

export function checkScope(input: ScopeInput): ScopeVerdict {
  if (input == null || typeof input !== "object") {
    return { result: "DENY", reason: "invalid input", code: "SCOPE_DENIED" };
  }

  const isWrite = input.isWrite === true;
  const isRead = input.isWrite === false;
  const targetPath = typeof input.targetPath === "string" ? input.targetPath : "";
  const declaredPaths = Array.isArray(input.declaredPaths) ? input.declaredPaths : [];
  if (!isWrite && !isRead) {
    return { result: "DENY", reason: "invalid isWrite field", code: "SCOPE_DENIED" };
  }
  if (!isWrite) return { result: "ALLOW" };
  if (!targetPath) {
    return { result: "DENY", reason: "unresolvable write target", code: "SCOPE_DENIED" };
  }

  const canonicalTarget = canonicalizePath(targetPath);
  for (const declared of declaredPaths) {
    if (typeof declared === "string" && isPathInScope(canonicalTarget, canonicalizePath(declared))) {
      return { result: "ALLOW" };
    }
  }
  return {
    result: "DENY",
    reason: `${canonicalTarget} not in declared paths [${declaredPaths.join(", ")}]`,
    code: "SCOPE_DENIED",
  };
}

function canonicalizePath(path: string): string {
  const resolved: string[] = [];
  for (const part of path.split("/")) {
    if (part === "." || part === "") continue;
    if (part === "..") resolved.pop();
    else resolved.push(part);
  }
  return resolved.join("/");
}

export function isPathInScope(target: string, scope: string): boolean {
  if (target === scope) return true;
  return scope.endsWith("/") ? target.startsWith(scope) : target.startsWith(`${scope}/`);
}

export type ScopeChangeKind = "create" | "modify" | "delete" | "rename";
export type ScopeOperationKind = "read" | ScopeChangeKind | "link";

export interface ScopeDeclaration {
  readonly path: string;
  readonly changeKind: ScopeChangeKind;
  readonly directory: boolean;
  readonly fromPath?: string;
}

export type ScopeOperation =
  | { readonly kind: "read" | "create" | "modify" | "delete"; readonly path: string; readonly directory: boolean }
  | { readonly kind: "rename" | "link"; readonly sourcePath: string; readonly destinationPath: string; readonly directory: boolean };

export interface StructuredScopeInput {
  /** Existing canonical worktree directory that bounds all authority. */
  readonly worktreeRoot: string;
  /** Existing canonical root from which every declaration and endpoint is resolved. */
  readonly authorizedRoot: string;
  /** Canonical state, attempt-policy/context, and bundle roots. */
  readonly reservedRoots: readonly string[];
  readonly declaredScope: readonly ScopeDeclaration[];
  readonly operation: ScopeOperation;
}

export type StructuredScopeVerdict =
  | { readonly result: "ALLOW"; readonly changeKind: ScopeChangeKind | null }
  | { readonly result: "ABSTAIN"; readonly changeKind: null }
  | { readonly result: "DENY"; readonly changeKind: ScopeChangeKind | null; readonly reason: string; readonly code: "RICKGENT_SCOPE_DENIED" };

interface ProvenDeclaration {
  readonly declaration: ScopeDeclaration;
  readonly path: string;
  readonly fromPath?: string;
}

class ScopeResolutionError extends Error {}

function deny(reason: string, changeKind: ScopeChangeKind | null = null): StructuredScopeVerdict {
  return { result: "DENY", changeKind, reason, code: "RICKGENT_SCOPE_DENIED" };
}

function pathInside(root: string, candidate: string): boolean {
  const rel = relative(root, candidate);
  return rel === "" || (rel !== ".." && !rel.startsWith(`..${sep}`) && !isAbsolute(rel));
}

function canonicalExistingDirectory(path: unknown, label: string): string {
  if (typeof path !== "string" || path === "" || !isAbsolute(path) || resolve(path) !== path) {
    throw new ScopeResolutionError(`${label} is not a canonical absolute path`);
  }
  let canonical: string;
  try {
    canonical = realpathSync(path);
    if (!statSync(canonical).isDirectory()) throw new ScopeResolutionError(`${label} is not a directory`);
  } catch (error) {
    if (error instanceof ScopeResolutionError) throw error;
    throw new ScopeResolutionError(`${label} is unavailable`);
  }
  if (canonical !== path) throw new ScopeResolutionError(`${label} is not canonical`);
  return canonical;
}

function validateRelativePath(path: unknown, label: string): string {
  if (typeof path !== "string" || path === "" || path.includes("\0")) {
    throw new ScopeResolutionError(`${label} is empty or malformed`);
  }
  if (
    isAbsolute(path)
    || /^[A-Za-z]:[\\/]/.test(path)
    || path.startsWith("\\\\")
    || path.includes("\\")
  ) {
    throw new ScopeResolutionError(`${label} must be a portable root-relative path`);
  }
  const components = path.split("/");
  if (components.some((part) => part === "" || part === "." || part === "..")) {
    throw new ScopeResolutionError(`${label} is not a canonical relative path`);
  }
  return path;
}

/** Resolve through the nearest existing parent without forgiving broken links or loops. */
function realpathNearestExisting(path: string): string {
  const normalized = resolve(path);
  let current = normalized;
  const tail: string[] = [];
  for (;;) {
    try {
      const canonical = realpathSync(current);
      return tail.length === 0 ? canonical : resolve(canonical, ...tail);
    } catch (error) {
      const code = (error as NodeJS.ErrnoException).code;
      if (code !== "ENOENT") throw new ScopeResolutionError("path cannot be resolved safely");
      try {
        lstatSync(current);
        // An existing object that realpath cannot resolve is a broken link.
        throw new ScopeResolutionError("path contains an unresolved symbolic link");
      } catch (probe) {
        if (probe instanceof ScopeResolutionError) throw probe;
        if ((probe as NodeJS.ErrnoException).code !== "ENOENT") {
          throw new ScopeResolutionError("path identity cannot be proven");
        }
      }
      const parent = dirname(current);
      if (parent === current) throw new ScopeResolutionError("path has no resolvable parent");
      tail.unshift(current.slice(parent.length + (parent.endsWith(sep) ? 0 : 1)));
      current = parent;
    }
  }
}

function lexicalGitPath(path: string): boolean {
  return path.split("/")[0] === ".git";
}

function overlapsReserved(path: string, directory: boolean, reservedRoots: readonly string[]): boolean {
  return reservedRoots.some((reserved) => pathInside(reserved, path) || (directory && pathInside(path, reserved)));
}

function endpointExists(path: string): boolean {
  try {
    lstatSync(path);
    return true;
  } catch (error) {
    if ((error as NodeJS.ErrnoException).code === "ENOENT") return false;
    throw new ScopeResolutionError("endpoint identity cannot be inspected");
  }
}

function declarationContains(declaration: ProvenDeclaration, endpoint: string, source = false): boolean {
  const declared = source ? declaration.fromPath : declaration.path;
  if (declared === undefined) return false;
  return declaration.declaration.directory ? pathInside(declared, endpoint) : declared === endpoint;
}

function validateDeclaration(value: ScopeDeclaration): void {
  if (value === null || typeof value !== "object") throw new ScopeResolutionError("scope declaration is malformed");
  if (!(["create", "modify", "delete", "rename"] as const).includes(value.changeKind) || typeof value.directory !== "boolean") {
    throw new ScopeResolutionError("scope declaration kind or directory flag is malformed");
  }
  validateRelativePath(value.path, "declared path");
  if (value.changeKind === "rename") validateRelativePath(value.fromPath, "declared rename source");
  else if (value.fromPath !== undefined) throw new ScopeResolutionError("non-rename declaration contains fromPath");
}

function proveEndpoint(
  rawPath: string,
  label: string,
  worktreeRoot: string,
  authorizedRoot: string,
  reservedRoots: readonly string[],
  directory: boolean,
): string {
  const relativePath = validateRelativePath(rawPath, label);
  if (lexicalGitPath(relativePath)) throw new ScopeResolutionError(`${label} enters .git`);
  const canonical = realpathNearestExisting(join(authorizedRoot, relativePath));
  if (!pathInside(worktreeRoot, canonical) || !pathInside(authorizedRoot, canonical)) {
    throw new ScopeResolutionError(`${label} resolves outside authorized scope`);
  }
  if (overlapsReserved(canonical, directory, reservedRoots)) {
    throw new ScopeResolutionError(`${label} overlaps a reserved root`);
  }
  return canonical;
}

function operationChangeKind(operation: ScopeOperation): ScopeChangeKind | null {
  if (operation.kind === "read") return null;
  return operation.kind === "link" ? "create" : operation.kind;
}

/**
 * Prove roots, every declaration, and every operation endpoint before applying
 * exact file/directory ownership and TicketContract change-kind semantics.
 */
export function checkStructuredScope(input: StructuredScopeInput): StructuredScopeVerdict {
  let observedKind: ScopeChangeKind | null = null;
  try {
    if (input == null || typeof input !== "object") return deny("scope request is malformed");
    const worktreeRoot = canonicalExistingDirectory(input.worktreeRoot, "worktree root");
    const authorizedRoot = canonicalExistingDirectory(input.authorizedRoot, "authorized root");
    if (!pathInside(worktreeRoot, authorizedRoot)) {
      return deny("authorized root is outside the canonical worktree");
    }
    if (!Array.isArray(input.reservedRoots) || !Array.isArray(input.declaredScope)) {
      return deny("reserved roots or declarations are malformed");
    }

    const reservedRoots = input.reservedRoots.map((root, index) => canonicalExistingDirectory(root, `reserved root ${index}`));
    const gitPath = join(worktreeRoot, ".git");
    try {
      reservedRoots.push(realpathSync(gitPath));
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code !== "ENOENT") return deny(".git identity cannot be proven");
    }

    const declarations: ProvenDeclaration[] = input.declaredScope.map((declaration) => {
      validateDeclaration(declaration);
      const path = proveEndpoint(
        declaration.path,
        "declared path",
        worktreeRoot,
        authorizedRoot,
        reservedRoots,
        declaration.directory,
      );
      if (declaration.changeKind !== "rename") return { declaration, path };
      const fromPath = proveEndpoint(
        declaration.fromPath!,
        "declared rename source",
        worktreeRoot,
        authorizedRoot,
        reservedRoots,
        declaration.directory,
      );
      return { declaration, path, fromPath };
    });

    const operation = input.operation;
    if (operation === null || typeof operation !== "object" || typeof operation.directory !== "boolean") {
      return deny("scope operation is malformed");
    }
    observedKind = operationChangeKind(operation);

    if (operation.kind === "read" || operation.kind === "create" || operation.kind === "modify" || operation.kind === "delete") {
      if ("sourcePath" in operation || "destinationPath" in operation) {
        return deny("single-endpoint operation contains extra endpoints", observedKind);
      }
      const endpoint = proveEndpoint(
        operation.path,
        "operation endpoint",
        worktreeRoot,
        authorizedRoot,
        reservedRoots,
        operation.directory,
      );
      const exists = endpointExists(join(authorizedRoot, validateRelativePath(operation.path, "operation endpoint")));
      if (operation.kind === "create" && exists) return deny("create endpoint already exists", observedKind);
      if ((operation.kind === "read" || operation.kind === "modify" || operation.kind === "delete") && !exists) {
        return deny(`${operation.kind} endpoint does not exist`, observedKind);
      }
      if (exists && statSync(endpoint).isDirectory() !== operation.directory) {
        return deny("operation endpoint type does not match its directory flag", observedKind);
      }
      const owners = declarations.filter((declaration) => declarationContains(declaration, endpoint));
      if (operation.kind === "read") {
        return owners.length > 0 ? { result: "ALLOW", changeKind: null } : { result: "ABSTAIN", changeKind: null };
      }
      if (!owners.some((owner) => owner.declaration.changeKind === operation.kind)) {
        return deny("operation path or change kind is outside declared scope", observedKind);
      }
      return { result: "ALLOW", changeKind: observedKind };
    }

    if (operation.kind !== "rename" && operation.kind !== "link") return deny("scope operation kind is unsupported");
    if ("path" in operation) return deny("two-endpoint operation contains path", observedKind);
    const source = proveEndpoint(
      operation.sourcePath,
      "operation source",
      worktreeRoot,
      authorizedRoot,
      reservedRoots,
      operation.directory,
    );
    const destination = proveEndpoint(
      operation.destinationPath,
      "operation destination",
      worktreeRoot,
      authorizedRoot,
      reservedRoots,
      operation.directory,
    );
    const sourceLexical = join(authorizedRoot, validateRelativePath(operation.sourcePath, "operation source"));
    const destinationLexical = join(authorizedRoot, validateRelativePath(operation.destinationPath, "operation destination"));
    if (!endpointExists(sourceLexical)) return deny(`${operation.kind} source does not exist`, observedKind);
    if (endpointExists(destinationLexical)) return deny(`${operation.kind} destination already exists`, observedKind);
    if (statSync(source).isDirectory() !== operation.directory) {
      return deny("operation source type does not match its directory flag", observedKind);
    }

    if (operation.kind === "rename") {
      const owned = declarations.some((declaration) =>
        declaration.declaration.changeKind === "rename"
        && declarationContains(declaration, source, true)
        && declarationContains(declaration, destination),
      );
      return owned
        ? { result: "ALLOW", changeKind: "rename" }
        : deny("rename endpoints do not match one declared rename", "rename");
    }

    const destinationOwned = declarations.some((declaration) =>
      declaration.declaration.changeKind === "create" && declarationContains(declaration, destination),
    );
    return destinationOwned
      ? { result: "ALLOW", changeKind: "create" }
      : deny("link destination is outside declared create scope", "create");
  } catch (error) {
    const reason = error instanceof ScopeResolutionError ? error.message : "scope identity could not be proven";
    return deny(reason, observedKind);
  }
}

// Compatibility shape retained for the verdict CLI. It delegates to the same
// structured engine and grants no native tool authority.
export interface ResolvedScopeInput {
  root: string;
  declaredPaths: string[];
  targetPath: string;
  isWrite: boolean;
  destinationPath?: string;
}

export function checkScopeResolved(input: ResolvedScopeInput): ScopeVerdict {
  if (input == null || typeof input !== "object") return { result: "DENY", reason: "invalid input", code: "SCOPE_DENIED" };
  if (input.isWrite === false) return { result: "ALLOW" };
  if (input.isWrite !== true) return { result: "DENY", reason: "invalid isWrite field", code: "SCOPE_DENIED" };
  if (!Array.isArray(input.declaredPaths) || typeof input.targetPath !== "string" || input.targetPath === "") {
    return { result: "DENY", reason: "unresolvable write target", code: "SCOPE_DENIED" };
  }
  const paths = input.declaredPaths.map((path) => typeof path === "string" ? path.replace(/\/+$/, "") : path as unknown as string);
  let operation: ScopeOperation;
  let declarations: ScopeDeclaration[];
  if (typeof input.destinationPath === "string" && input.destinationPath !== "") {
    operation = { kind: "rename", sourcePath: input.targetPath, destinationPath: input.destinationPath, directory: false };
    declarations = paths.map((path) => ({ path, fromPath: path, changeKind: "rename", directory: true }));
  } else {
    let exists = false;
    try {
      exists = endpointExists(join(input.root, validateRelativePath(input.targetPath, "operation endpoint")));
    } catch {
      // The structured engine will return the canonical denial.
    }
    const kind: ScopeChangeKind = exists ? "modify" : "create";
    operation = { kind, path: input.targetPath, directory: false };
    declarations = paths.map((path) => ({ path, changeKind: kind, directory: true }));
  }
  const verdict = checkStructuredScope({
    worktreeRoot: input.root,
    authorizedRoot: input.root,
    reservedRoots: [],
    declaredScope: declarations,
    operation,
  });
  return verdict.result === "DENY"
    ? { result: "DENY", reason: verdict.reason, code: "SCOPE_DENIED" }
    : { result: "ALLOW" };
}

File: rickgent-policies/rickgent_policies/scope.py
"""Canonical Rickgent scope authority and native FunctionPolicy handler.

Only :class:`CanonicalPolicyEvent` values produced from an authenticated attempt
context reach the scope engine. Raw event/config aliases and shell command text
never grant authority.
"""

from __future__ import annotations

import os
import re
import stat
from dataclasses import dataclass
from typing import Literal

from .context import FilesystemContextAuthenticator
from .policy_event import (
    CanonicalPolicyEvent,
    PolicyAbstention,
    PolicyDenial,
    TicketScopeEntry,
    adapt_native_policy_event,
)


SCOPE_DENIAL_CODE = "RICKGENT_SCOPE_DENIED"
CANONICAL_FILESYSTEM_TOOLS = frozenset(
    {"sys_os_read", "sys_os_write", "sys_os_edit"}
)
RAW_SHELL_TOOLS = frozenset({"sys_os_shell", "Bash", "bash", "Shell", "shell"})

ScopeChangeKind = Literal["create", "modify", "delete", "rename"]
ScopeOperationKind = Literal["read", "create", "modify", "delete", "rename", "link"]
ScopeDisposition = Literal["ALLOW", "DENY", "ABSTAIN"]


@dataclass(frozen=True)
class ScopeOperation:
    kind: ScopeOperationKind
    directory: bool
    path: str | None = None
    source_path: str | None = None
    destination_path: str | None = None


@dataclass(frozen=True)
class ScopeDecision:
    result: ScopeDisposition
    change_kind: ScopeChangeKind | None = None
    reason: str | None = None
    code: str | None = None


@dataclass(frozen=True)
class _ProvenDeclaration:
    declaration: TicketScopeEntry
    path: str
    from_path: str | None


class _ScopeError(Exception):
    pass


def _deny(reason: str, change_kind: ScopeChangeKind | None = None) -> ScopeDecision:
    return ScopeDecision("DENY", change_kind, reason, SCOPE_DENIAL_CODE)


def _path_inside(root: str, candidate: str) -> bool:
    try:
        return os.path.commonpath((root, candidate)) == root
    except (TypeError, ValueError):
        return False


def _canonical_existing_directory(value: object, label: str) -> str:
    if (
        not isinstance(value, str)
        or not value
        or not os.path.isabs(value)
        or os.path.normpath(value) != value
    ):
        raise _ScopeError(f"{label} is not a canonical absolute path")
    try:
        canonical = os.path.realpath(value, strict=True)
        info = os.stat(canonical)
    except (FileNotFoundError, NotADirectoryError, PermissionError, OSError):
        raise _ScopeError(f"{label} is unavailable") from None
    if canonical != value:
        raise _ScopeError(f"{label} is not canonical")
    if not stat.S_ISDIR(info.st_mode):
        raise _ScopeError(f"{label} is not a directory")
    return canonical


def _validate_relative_path(value: object, label: str) -> str:
    if not isinstance(value, str) or not value or "\0" in value:
        raise _ScopeError(f"{label} is empty or malformed")
    if (
        os.path.isabs(value)
        or re.match(r"^[A-Za-z]:[\\/]", value)
        or value.startswith("\\\\")
        or "\\" in value
    ):
        raise _ScopeError(f"{label} must be a portable root-relative path")
    components = value.split("/")
    if any(component in {"", ".", ".."} for component in components):
        raise _ScopeError(f"{label} is not a canonical relative path")
    return value


def _realpath_nearest_existing(path: str) -> str:
    current = os.path.abspath(path)
    tail: list[str] = []
    while True:
        try:
            canonical = os.path.realpath(current, strict=True)
            return os.path.abspath(os.path.join(canonical, *tail))
        except FileNotFoundError:
            # lexists distinguishes an absent component from a dangling link.
            if os.path.lexists(current):
                raise _ScopeError("path contains an unresolved symbolic link") from None
            parent, name = os.path.split(current)
            if parent == current or not name:
                raise _ScopeError("path has no resolvable parent") from None
            tail.insert(0, name)
            current = parent
        except (NotADirectoryError, PermissionError, OSError):
            raise _ScopeError("path cannot be resolved safely") from None


def _lexical_git_path(path: str) -> bool:
    return path.split("/", 1)[0] == ".git"


def _overlaps_reserved(
    path: str, directory: bool, reserved_roots: tuple[str, ...]
) -> bool:
    return any(
        _path_inside(reserved, path)
        or (directory and _path_inside(path, reserved))
        for reserved in reserved_roots
    )


def _prove_endpoint(
    raw_path: str,
    label: str,
    *,
    worktree_root: str,
    authorized_root: str,
    reserved_roots: tuple[str, ...],
    directory: bool,
) -> str:
    relative_path = _validate_relative_path(raw_path, label)
    if _lexical_git_path(relative_path):
        raise _ScopeError(f"{label} enters .git")
    canonical = _realpath_nearest_existing(
        os.path.join(authorized_root, relative_path)
    )
    if not _path_inside(worktree_root, canonical) or not _path_inside(
        authorized_root, canonical
    ):
        raise _ScopeError(f"{label} resolves outside authorized scope")
    if _overlaps_reserved(canonical, directory, reserved_roots):
        raise _ScopeError(f"{label} overlaps a reserved root")
    return canonical


def _endpoint_exists(path: str) -> bool:
    try:
        os.lstat(path)
        return True
    except FileNotFoundError:
        return False
    except OSError:
        raise _ScopeError("endpoint identity cannot be inspected") from None


def _declaration_contains(
    declaration: _ProvenDeclaration, endpoint: str, *, source: bool = False
) -> bool:
    declared = declaration.from_path if source else declaration.path
    if declared is None:
        return False
    if declaration.declaration.directory:
        return _path_inside(declared, endpoint)
    return declared == endpoint


def _validate_declaration(declaration: object) -> TicketScopeEntry:
    if not isinstance(declaration, TicketScopeEntry):
        raise _ScopeError("scope declaration is malformed")
    _validate_relative_path(declaration.path, "declared path")
    if declaration.change_kind not in {"create", "modify", "delete", "rename"}:
        raise _ScopeError("scope declaration change kind is malformed")
    if not isinstance(declaration.directory, bool):
        raise _ScopeError("scope declaration directory flag is malformed")
    if declaration.change_kind == "rename":
        _validate_relative_path(declaration.from_path, "declared rename source")
    elif declaration.from_path is not None:
        raise _ScopeError("non-rename declaration contains from_path")
    return declaration


def _operation_change_kind(operation: ScopeOperation) -> ScopeChangeKind | None:
    if operation.kind == "read":
        return None
    if operation.kind == "link":
        return "create"
    return operation.kind


def evaluate_scope(
    *,
    worktree_root: str,
    authorized_root: str,
    reserved_roots: tuple[str, ...],
    declared_scope: tuple[TicketScopeEntry, ...],
    operation: ScopeOperation,
) -> ScopeDecision:
    """Evaluate one closed structured filesystem request."""

    observed_kind: ScopeChangeKind | None = None
    try:
        canonical_worktree = _canonical_existing_directory(
            worktree_root, "worktree root"
        )
        canonical_authorized = _canonical_existing_directory(
            authorized_root, "authorized root"
        )
        if not _path_inside(canonical_worktree, canonical_authorized):
            return _deny("authorized root is outside the canonical worktree")
        if not isinstance(reserved_roots, tuple) or not isinstance(
            declared_scope, tuple
        ):
            return _deny("reserved roots or declarations are malformed")

        proven_reserved = tuple(
            _canonical_existing_directory(root, f"reserved root {index}")
            for index, root in enumerate(reserved_roots)
        )
        git_path = os.path.join(canonical_worktree, ".git")
        if os.path.lexists(git_path):
            try:
                proven_reserved += (os.path.realpath(git_path, strict=True),)
            except OSError:
                return _deny(".git identity cannot be proven")

        declarations: list[_ProvenDeclaration] = []
        for raw_declaration in declared_scope:
            declaration = _validate_declaration(raw_declaration)
            path = _prove_endpoint(
                declaration.path,
                "declared path",
                worktree_root=canonical_worktree,
                authorized_root=canonical_authorized,
                reserved_roots=proven_reserved,
                directory=declaration.directory,
            )
            from_path = None
            if declaration.change_kind == "rename":
                assert declaration.from_path is not None
                from_path = _prove_endpoint(
                    declaration.from_path,
                    "declared rename source",
                    worktree_root=canonical_worktree,
                    authorized_root=canonical_authorized,
                    reserved_roots=proven_reserved,
                    directory=declaration.directory,
                )
            declarations.append(_ProvenDeclaration(declaration, path, from_path))

        if not isinstance(operation, ScopeOperation) or not isinstance(
            operation.directory, bool
        ):
            return _deny("scope operation is malformed")
        observed_kind = _operation_change_kind(operation)

        if operation.kind in {"read", "create", "modify", "delete"}:
            if operation.source_path is not None or operation.destination_path is not None:
                return _deny("single-endpoint operation contains extra endpoints", observed_kind)
            endpoint_path = _validate_relative_path(
                operation.path, "operation endpoint"
            )
            endpoint = _prove_endpoint(
                endpoint_path,
                "operation endpoint",
                worktree_root=canonical_worktree,
                authorized_root=canonical_authorized,
                reserved_roots=proven_reserved,
                directory=operation.directory,
            )
            exists = _endpoint_exists(
                os.path.join(canonical_authorized, endpoint_path)
            )
            if operation.kind == "create" and exists:
                return _deny("create endpoint already exists", observed_kind)
            if operation.kind in {"read", "modify", "delete"} and not exists:
                return _deny(
                    f"{operation.kind} endpoint does not exist", observed_kind
                )
            if exists and stat.S_ISDIR(os.stat(endpoint).st_mode) != operation.directory:
                return _deny(
                    "operation endpoint type does not match its directory flag",
                    observed_kind,
                )
            owners = [
                declaration
                for declaration in declarations
                if _declaration_contains(declaration, endpoint)
            ]
            if operation.kind == "read":
                if owners:
                    return ScopeDecision("ALLOW")
                return ScopeDecision("ABSTAIN")
            if not any(
                owner.declaration.change_kind == operation.kind for owner in owners
            ):
                return _deny(
                    "operation path or change kind is outside declared scope",
                    observed_kind,
                )
            return ScopeDecision("ALLOW", observed_kind)

        if operation.kind not in {"rename", "link"}:
            return _deny("scope operation kind is unsupported")
        if operation.path is not None:
            return _deny("two-endpoint operation contains path", observed_kind)
        source_path = _validate_relative_path(operation.source_path, "operation source")
        destination_path = _validate_relative_path(
            operation.destination_path, "operation destination"
        )
        source = _prove_endpoint(
            source_path,
            "operation source",
            worktree_root=canonical_worktree,
            authorized_root=canonical_authorized,
            reserved_roots=proven_reserved,
            directory=operation.directory,
        )
        destination = _prove_endpoint(
            destination_path,
            "operation destination",
            worktree_root=canonical_worktree,
            authorized_root=canonical_authorized,
            reserved_roots=proven_reserved,
            directory=operation.directory,
        )
        if not _endpoint_exists(os.path.join(canonical_authorized, source_path)):
            return _deny(f"{operation.kind} source does not exist", observed_kind)
        if _endpoint_exists(os.path.join(canonical_authorized, destination_path)):
            return _deny(
                f"{operation.kind} destination already exists", observed_kind
            )
        if stat.S_ISDIR(os.stat(source).st_mode) != operation.directory:
            return _deny(
                "operation source type does not match its directory flag",
                observed_kind,
            )

        if operation.kind == "rename":
            owned = any(
                declaration.declaration.change_kind == "rename"
                and _declaration_contains(declaration, source, source=True)
                and _declaration_contains(declaration, destination)
                for declaration in declarations
            )
            if owned:
                return ScopeDecision("ALLOW", "rename")
            return _deny(
                "rename endpoints do not match one declared rename", "rename"
            )

        destination_owned = any(
            declaration.declaration.change_kind == "create"
            and _declaration_contains(declaration, destination)
            for declaration in declarations
        )
        if destination_owned:
            return ScopeDecision("ALLOW", "create")
        return _deny("link destination is outside declared create scope", "create")
    except _ScopeError as error:
        return _deny(str(error), observed_kind)
    except Exception:
        return _deny("scope identity could not be proven", observed_kind)


def _native_operation(event: CanonicalPolicyEvent) -> ScopeOperation | ScopeDecision:
    path = event.source_endpoint
    if event.action == "read":
        return ScopeOperation("read", False, path=path)
    if event.action == "write":
        try:
            relative_path = _validate_relative_path(path, "operation endpoint")
            exists = _endpoint_exists(
                os.path.join(event.worktree_realpath, relative_path)
            )
        except _ScopeError as error:
            return _deny(str(error))
        return ScopeOperation("modify" if exists else "create", False, path=path)
    if event.action == "edit":
        return ScopeOperation("modify", False, path=path)
    return _deny("canonical native action is unsupported")


def evaluate_canonical_event(event: CanonicalPolicyEvent) -> ScopeDecision:
    """Project an authenticated canonical native event into scope semantics."""

    if event.native_phase != "tool_call":
        return ScopeDecision("ABSTAIN")
    operation = _native_operation(event)
    if isinstance(operation, ScopeDecision):
        return operation
    return evaluate_scope(
        worktree_root=event.worktree_realpath,
        authorized_root=event.worktree_realpath,
        reserved_roots=(
            event.state_root_realpath,
            event.policy_root_realpath,
            event.bundle_root_realpath,
        ),
        declared_scope=event.declared_scope,
        operation=operation,
    )


def _policy_mapping(decision: ScopeDecision) -> dict[str, str] | None:
    if decision.result == "ABSTAIN":
        return None
    if decision.result == "ALLOW":
        return {"result": "ALLOW"}
    reason = decision.reason or "scope request denied"
    return {
        "result": "DENY",
        "reason": f"{SCOPE_DENIAL_CODE}: {reason}",
        "code": SCOPE_DENIAL_CODE,
    }


def scope_fence(event: object, config: object) -> dict[str, str] | None:
    """Real FunctionPolicy entrypoint backed only by canonical authority."""

    try:
        outcome = adapt_native_policy_event(
            event,
            config,
            authenticator=FilesystemContextAuthenticator.from_environment(),
        )
        if isinstance(outcome, PolicyDenial):
            return {
                "result": "DENY",
                "reason": outcome.reason,
                "code": outcome.code.value,
            }
        if isinstance(outcome, PolicyAbstention):
            return None
        return _policy_mapping(evaluate_canonical_event(outcome))
    except Exception:
        return {
            "result": "DENY",
            "reason": f"{SCOPE_DENIAL_CODE}: scope policy failed safely",
            "code": SCOPE_DENIAL_CODE,
        }


def check_scope_resolved(
    root: object,
    declared_paths: object,
    target_path: object,
    is_write: object,
    destination_path: object = None,
) -> dict[str, str]:
    """Compatibility adapter for the legacy verdict CLI shape.

    It delegates to :func:`evaluate_scope`; it is not used by the production
    FunctionPolicy handler and cannot supply native authority.
    """

    if is_write is False:
        return {"result": "ALLOW"}
    if is_write is not True or not isinstance(root, str):
        return {
            "result": "DENY",
            "reason": "invalid legacy scope request",
            "code": "SCOPE_DENIED",
        }
    if not isinstance(declared_paths, (list, tuple)):
        declared_paths = ()
    normalized = tuple(
        path.rstrip("/") if isinstance(path, str) else path
        for path in declared_paths
    )
    try:
        target = _validate_relative_path(target_path, "operation endpoint")
        exists = _endpoint_exists(os.path.join(root, target))
    except _ScopeError:
        exists = False

    if isinstance(destination_path, str) and destination_path:
        operation = ScopeOperation(
            "rename",
            False,
            source_path=target_path if isinstance(target_path, str) else None,
            destination_path=destination_path,
        )
        declarations = tuple(
            TicketScopeEntry(path, "rename", True, path)
            for path in normalized
            if isinstance(path, str)
        )
    else:
        kind: ScopeChangeKind = "modify" if exists else "create"
        operation = ScopeOperation(
            kind,
            False,
            path=target_path if isinstance(target_path, str) else None,
        )
        declarations = tuple(
            TicketScopeEntry(path, kind, True)
            for path in normalized
            if isinstance(path, str)
        )
    decision = evaluate_scope(
        worktree_root=root,
        authorized_root=root,
        reserved_roots=(),
        declared_scope=declarations,
        operation=operation,
    )
    if decision.result == "DENY":
        return {
            "result": "DENY",
            "reason": decision.reason or "scope request denied",
            "code": "SCOPE_DENIED",
        }
    return {"result": "ALLOW"}


__all__ = [
    "CANONICAL_FILESYSTEM_TOOLS",
    "RAW_SHELL_TOOLS",
    "SCOPE_DENIAL_CODE",
    "ScopeDecision",
    "ScopeOperation",
    "check_scope_resolved",
    "evaluate_canonical_event",
    "evaluate_scope",
    "scope_fence",
]
